File: LeetCode/Medium/0120-triangle/0120-triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        next_value = triangle[-1][:]
        current_value = [0] * n
        for i in range(n - 2, -1, -1):
            for j in range(i + 1):
                right_value = triangle[i][j] + next_value[j]
                left_value = triangle[i][j] + next_value[j + 1]
                current_value[j] = min(right_value, left_value)
            current_value, next_value = next_value, current_value
        return next_value[0]

File: 0120-triangle/0120-triangle.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int minimumTotal(vector<vector<int>> &triangle)
    {
        vector<int> tabulation = triangle[triangle.size() - 1];
        for (int i = triangle.size() - 2; i >= 0; --i) 
        {
            for (int j = 0; j < triangle[i].size(); ++j)  
            {
                tabulation[j] = triangle[i][j] + min(tabulation[j], tabulation[j + 1]);
            }
        }
        
        return tabulation[0];
    }
};